### PR TITLE
Avoid C-style cast in fe

### PR DESCRIPTION
**Pull request by masterleinad**
Part of #7459. 
This PR replaces all C-style casts in `include/deal.II/fe` and `source/fe`.

### CODE REVIEW

File: source/fe/fe_poly.cc [hunk 1]
```diff
@@ -43,13 +43,9 @@ FE_Poly<TensorProductPolynomials<1>, 1, 2>::fill_fe_values(
...
-  // convert data object to internal
-  // data for this class. fails with
-  // an exception if that is not
-  // possible
-  Assert(dynamic_cast<const InternalData *>(&fe_internal) != nullptr,
-         ExcInternalError());
-  const InternalData &fe_data = static_cast<const InternalData &>(fe_internal);
+  // convert data object to internal data for this class. fails with an
+  // exception if that is not possible
+  const InternalData &fe_data = dynamic_cast<const InternalData &>(fe_internal);
```
Comment (bangerth) [inline source/fe/fe_poly.cc]: Please leave the assertion in place. It's true that the `dynamic_cast` fails, but with a `std::bad_cast` exception that doesn't reveal anything about the location where the problem happened.

Same for the cases below.

Comment (masterleinad) [reply source/fe/fe_poly.cc]: I was quite sure that there would some kind of discussion here.

Comment (bangerth) [reply source/fe/fe_poly.cc]: :-)

Comment (masterleinad) [reply source/fe/fe_poly.cc]: As usual for all these PRs, I am happy to incorporate any comments.

**Pull request merged**